duckduckgo/Android
# Allow users to recover from WebView renderer failures

Task/Issue URL: https://app.asana.com/0/0/1152512237237203/f
Tech Design URL: https://app.asana.com/0/0/1151307893053517/f
CC: 

### **Description**:
Introducing gracefully handling WebView crash errors into our Browser.
A WebView render process can be gone due to a rendering crash or memory pressure situations. When we receive such event from WebViewClient we will avoid crashing, and instead, we will show a snackbar to inform about what happened and the possibility to reload the current URL. Also, application will still be usable so users will be able to perform other actions as closing the tab or navigate to a new URL.

### **Steps to test this PR**:
As we can't reproduce easily memory pressure situations or crashes, we will need to change our source code a bit in order to manually produce a crash.
Easiest way is to call `recoverFromRenderProcessGone()` when user clicks in a menu option, example will be inside `onBrokenSiteSelected()`.

### **Scenarios to test**
***Scenario to test***:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Verify user can't navigate back/forward/find a page/report a site

***Scenario to test***:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Verify snackbar appears
- Reload from snackbar
- Verify current tab is closed and a new one is opened with same URL

**Scenario to test**:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Verify snackbar appears
- Reload from option menu
- Verify current tab is closed and a new one is opened with same URL

**Scenario to test**:
- Open app
- submit an URL
- Navigate to produce some history navigation
- crash the app (with hardcoded code)
- Verify snackbar appears
- press back
- Verify home is shown
- press back
- Verify exits the app

**Scenario to test**:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Verify snackbar appears
- Submit current or a new URL
- Verify current tab is closed and a new one is opened with expected URL

**Scenario to test**:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Verify snackbar appears
- Open privacy settings and disable or enable toogle
- Go back to the tab
- Verify current tab is closed and a new one is opened with same URL

**Scenario to test**:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Open a new tab
- submit an URL
- crash the app (with hardcoded code)
- Change between tabs, snackbar should appear

**Scenario to test**:
- Open app
- submit an URL
- crash the app (with hardcoded code)
- Open a new tab
- submit an URL
- crash the app (with hardcoded code)
- Send app to background
- Bring back to foreground the app
- Snakbar should appear

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
